File: 0106-construct-binary-tree-from-inorder-and-postorder-traversal/0106-construct-binary-tree-from-inorder-and-postorder-traversal.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
use std::collections::HashMap;

impl Solution {
    pub fn build_tree(inorder: Vec<i32>, postorder: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
        
        let mut in_map : HashMap<i32, usize> = HashMap::new();
        
        for(idx, &val)  in inorder.iter().enumerate(){
            in_map.insert(val, idx);
        }
        
        let l_post =0;
        let r_post = postorder.len() as i32 -1;
        let l_in = 0;
        let r_in = postorder.len() as i32 -1;
        
        
        Solution::build(&postorder, l_post, r_post, &in_map, &inorder, l_in, r_in)
    }
    
    
    fn build(postorder: &Vec<i32>, l_post:i32, r_post: i32,
                        in_map: &HashMap<i32, usize>,
            inorder: &Vec<i32>, l_in: i32, r_in: i32) -> Option <Rc<RefCell<TreeNode>>>{
        
                
                if l_post > r_post || l_in > r_in{
                    return None;
                }
                
                let root_ele = postorder[r_post as usize];
                
                if let Some(&root_ele_idx) = in_map.get(&root_ele) {
                    
                    let noe = root_ele_idx as i32 - l_in;
                    
                    let mut root = TreeNode::new(root_ele);
                    
                    root.left = Solution::build(postorder, l_post, l_post+noe-1, in_map, inorder, l_in, root_ele_idx as i32-1);
                    root.right = Solution::build(postorder, l_post+noe, r_post-1, in_map, inorder, root_ele_idx as i32+1, r_in );
                    
                    
                    Some(Rc::new(RefCell::new(root)))
                }else{
                    None
                }
    }
}

File: 0106-construct-binary-tree-from-inorder-and-postorder-traversal/0106-construct-binary-tree-from-inorder-and-postorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    
        TreeNode* build(vector<int>& postorder, int l_post, int r_post, vector<int>& inorder, int l_in, int r_in, unordered_map<int, int> &inMap){
        
        //Valid Range
        if(l_post >r_post || l_in >r_in)
            return NULL;
        
        int rootEle = postorder[r_post];
        
        int rootEleIdx = inMap[rootEle]; 
            
        //noe on the left hand subtree of the node
        int noe = rootEleIdx-l_in;      // [l_in       ] rootEleIdx [.........r_in]

        TreeNode *root = new TreeNode(rootEle);
        root->left = build(postorder, l_post, l_post+noe-1, inorder, l_in, rootEleIdx-1, inMap);
        root->right = build(postorder, l_post+noe, r_post-1, inorder, rootEleIdx+1, r_in, inMap);
            
            
        return root;
    }
    
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        
        unordered_map<int, int>inMap;
        for(int idx =0; idx < inorder.size(); idx++)
            inMap.insert({inorder[idx], idx});
        
        int l_post =0, r_post = postorder.size()-1;
        int l_in = 0, r_in = inorder.size()-1;
        
        
        
        TreeNode * root = build(postorder, l_post, r_post, inorder, l_in, r_in, inMap);
        return root;
    }
};